### PR TITLE
cpu/atmega2560: reworked UART driver

### PR DESCRIPTION
**Pull request by haukepetersen**
gets rid of unneccesary configuration stuff in the boards periph_conf, makes the UART configuration much easier and saves  ~650 byte flash for the default example

### CODE REVIEW

File: cpu/atmega2560/include/atmega2560_regs.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * @brief    Base register address definitions
+ * @{
+ */
+#define MEGA_UART0_BASE         (uint16_t *)(&UCSR0A)
+#define MEGA_UART1_BASE         (uint16_t *)(&UCSR1A)
+#define MEGA_UART2_BASE         (uint16_t *)(&UCSR2A)
+#define MEGA_UART3_BASE         (uint16_t *)(&UCSR2A)
+/** @} */
```
Comment (kYc0o) [inline cpu/atmega2560/include/atmega2560_regs.h]: No extra parentheses required here?

File: cpu/atmega2560/periph/uart.c [hunk 2]
```diff
@@ -19,216 +19,115 @@
...
-            UART3_BAUD_RATE_H = (clock_divider >> 8);
-            break;
-#endif /* UART_3 */
-        default:
-            (void)clock_divider;    /* this makes cppcheck happy */
-    }
+    /* configure UART to 8N1 mode */
+    dev[uart]->CSRC = (1 << UCSZ00) | (1 << UCSZ01);
+    /* set clock divider */
+    dev[uart]->BRR = CLOCK_CORECLOCK / (16 * baudrate);
```
Comment (kYc0o) [inline cpu/atmega2560/periph/uart.c]: The meaning of this "16" is the CPU clock frequency?

Comment (haukepetersen) [inline cpu/atmega2560/periph/uart.c]: nope, UART oversamples 16 times -> the values comes from the datasheet

File: cpu/atmega2560/include/atmega2560_regs.h (hunk 1 above)
Comment (haukepetersen) [inline cpu/atmega2560/include/atmega2560_regs.h]: no really, but they won't hurt either -> will fix

File: cpu/atmega2560/periph/uart.c (hunk 2 above)
Comment (kYc0o) [inline cpu/atmega2560/periph/uart.c]: Ok, it's just that it looks like a magic number.

**Comment by kYc0o** [comment]
can you rebase since #5014 was merged? :)

**Comment by haukepetersen** [comment]
yepp

**Comment by haukepetersen** [comment]
rebased

**Comment by haukepetersen** [comment]
also added the missing parens

**Comment by kYc0o** [comment]
great thanks

**Comment by kYc0o** [comment]
is this normal?:
```
UART used for STDIO (the shell): UART_DEV(-1)
```

**Comment by kYc0o** [comment]
I'm trying to use tests/periph_uart and it freezes. After a reboot it prints trash, lots of trash :S

**Comment by kYc0o** [comment]
actually it is possible to initialise one UART but at the second one it freezes

File: cpu/atmega2560/periph/uart.c (hunk 2 above)
Comment (kYc0o) [inline cpu/atmega2560/periph/uart.c]: Uhm... maybe I'm misunderstanding the manual but... it shouldn't be
```
dev[uart]->BRR = (CLOCK_CORECLOCK / 16 / baudrate - 1)
```

On the manual it says:
```
#define FOSC 1843200// Clock Speed
#define BAUD 9600
#define (MYUBRR FOSC/16/BAUD-1)
void main( void )
{...
USART_Init ( MYUBRR );
...} // main
void USART_Init( unsigned int ubrr){
/* Set baud rate */
UBRRH = (unsigned char)(ubrr>>8);
UBRRL = (unsigned char)ubrr;
/* Enable receiver and transmitter */
UCSRB = (1<<RXEN)|(1<<TXEN);
/* Set frame format: 8data, 2stop bit */
UCSRC = (1<<USBS)|(3<<UCSZ0);
} // USART_Init
```

**Comment by haukepetersen** [comment]
still looking into this, but won't probably make it for the release, so putting this on the list for the next one...

**Comment by kYc0o** [comment]
seems reasonable :)

**Comment by PeterKietzmann** [comment]
#5537 was merged

**Pull request closed without merging**